File: android/app/src/main/kotlin/app/donottype/core/PersonalDictionary.kt
package app.donottype.core

import org.json.JSONArray
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets

/** A bounded spelling reference entered by the user or learned from an opted-in correction. */
object PersonalDictionary {
    const val MAX_TERMS = 100
    const val MAX_CHARACTERS_PER_TERM = 50

    class ValidationException(message: String) : IllegalArgumentException(message)

    fun normalize(raw: String): String {
        if (raw.any { it == '\r' || it == '\n' }) {
            throw ValidationException("A dictionary entry must fit on one line.")
        }
        val term = raw.trim().split(Regex("\\s+")).filter(String::isNotEmpty).joinToString(" ")
        if (term.isEmpty()) throw ValidationException("Enter a word or phrase.")
        if (term.codePointCount(0, term.length) > MAX_CHARACTERS_PER_TERM) {
            throw ValidationException(
                "Dictionary entries can be at most $MAX_CHARACTERS_PER_TERM characters.",
            )
        }
        return term
    }

    fun sanitize(raw: Iterable<String>?): List<String> {
        val seen = mutableSetOf<String>()
        return buildList {
            for (value in raw ?: emptyList()) {
                val term = runCatching { normalize(value) }.getOrNull() ?: continue
                if (!seen.add(term.lowercase())) continue
                add(term)
                if (size == MAX_TERMS) break
            }
        }
    }

    fun add(raw: String, terms: Iterable<String>): List<String> {
        val current = sanitize(terms)
        if (current.size >= MAX_TERMS) {
            throw ValidationException("The dictionary can contain at most $MAX_TERMS entries.")
        }
        val term = normalize(raw)
        if (current.any { it.equals(term, ignoreCase = true) }) {
            throw ValidationException("“$term” is already in the dictionary.")
        }
        return current + term
    }

    fun entriesFromCsv(data: ByteArray): List<String> {
        val decoder = StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
        var text = try {
            decoder.decode(ByteBuffer.wrap(data)).toString()
        } catch (_: Exception) {
            throw ValidationException("The file is not UTF-8 text.")
        }
        text = text.removePrefix("\uFEFF").replace("\r\n", "\n").replace('\r', '\n')
        val seen = mutableSetOf<String>()
        return buildList {
            text.split('\n').forEachIndexed { index, line ->
                if (line.isBlank()) return@forEachIndexed
                val term = normalize(csvField(line, index + 1))
                if (!seen.add(term.lowercase())) return@forEachIndexed
                add(term)
                if (size > MAX_TERMS) {
                    throw ValidationException("The dictionary can contain at most $MAX_TERMS entries.")
                }
            }
        }
    }

    fun referenceBlock(raw: Iterable<String>): String? {
        val terms = sanitize(raw)
        if (terms.isEmpty()) return null
        return """
            PERSONAL DICTIONARY — SPELLING REFERENCE ONLY, DO NOT TRANSCRIBE
            The user supplied the JSON strings below as possible spellings. Use an entry only when
            the same word or phrase is audible. The list is not evidence that an entry was spoken,
            and it never overrides clear audio. Digits, versions and quantities come from audio
            alone even when an entry contains a number.
            ${JSONArray(terms)}
            END PERSONAL DICTIONARY. The audio is still the ONLY thing to transcribe.
        """.trimIndent()
    }

    fun keyterms(raw: Iterable<String>, maxTerms: Int, maxCharsPerTerm: Int): List<String> {
        if (maxTerms <= 0) return emptyList()
        return sanitize(raw).filter {
            it.none(Char::isDigit) && it.codePointCount(0, it.length) <= maxCharsPerTerm
        }.take(maxTerms)
    }

    fun mergeKeyterms(
        dictionary: Iterable<String>,
        derived: Iterable<String>,
        maxTerms: Int,
        maxCharsPerTerm: Int,
    ): List<String> {
        val result = keyterms(dictionary, maxTerms, maxCharsPerTerm).toMutableList()
        val seen = result.mapTo(mutableSetOf()) { it.lowercase() }
        for (term in derived) {
            if (term.codePointCount(0, term.length) > maxCharsPerTerm) continue
            if (!seen.add(term.lowercase())) continue
            result.add(term)
            if (result.size == maxTerms) break
        }
        return result
    }

    /** Returns only spelling/capitalisation fixes, never insertions, deletions or number changes. */
    fun learnedCandidates(original: String, corrected: String): List<String> {
        if (original == corrected) return emptyList()
        val left = tokenize(original)
        val right = tokenize(corrected)
        val candidates = mutableListOf<String>()
        for ((before, after) in alignedDifferences(left, right)) {
            if (classify(before, after) == Difference.SPELLING_FIXED) {
                usableLearnedTerm(after.joinToString(" "))?.let(candidates::add)
            } else {
                bestSpellingSubspan(before, after)?.let(candidates::add)
            }
        }
        if (left.size == right.size) {
            left.zip(right).forEach { (before, after) ->
                if (normalizeToken(before) == normalizeToken(after) && before != after &&
                    before.equals(after, ignoreCase = true)
                ) {
                    usableLearnedTerm(after)?.let(candidates::add)
                }
            }
        }
        return sanitize(candidates)
    }

    private fun csvField(line: String, lineNumber: Int): String {
        val trimmed = line.trim()
        if (!trimmed.startsWith('"')) {
            if (line.contains(',')) throw ValidationException(
                "Line $lineNumber has more than one CSV column. Use one entry per row.",
            )
            return line
        }
        val value = StringBuilder()
        var index = 1
        var closed = false
        while (index < trimmed.length) {
            if (trimmed[index] == '"') {
                if (index + 1 < trimmed.length && trimmed[index + 1] == '"') {
                    value.append('"'); index += 2; continue
                }
                closed = true; index++; break
            }
            value.append(trimmed[index++])
        }
        if (!closed) throw ValidationException(
            "Line $lineNumber has an unterminated or malformed quoted value.",
        )
        val remainder = trimmed.substring(index).trim()
        if (remainder.isNotEmpty()) {
            if (remainder.startsWith(',')) throw ValidationException(
                "Line $lineNumber has more than one CSV column. Use one entry per row.",
            )
            throw ValidationException(
                "Line $lineNumber has an unterminated or malformed quoted value.",
            )
        }
        return value.toString()
    }

    private enum class Difference { SPELLING_FIXED, CONTENT_CHANGED }
    private data class Span(val left: List<String>, val right: List<String>)

    private fun tokenize(text: String): List<String> = text.split(Regex("\\s+"))
        .filter { normalizeToken(it).isNotEmpty() }

    private fun normalizeToken(text: String): String =
        text.lowercase().filter { it.isLetterOrDigit() }

    private fun digitRuns(text: String): List<String> = Regex("\\d+").findAll(text).map { it.value }.toList()

    private fun classify(left: List<String>, right: List<String>): Difference {
        if (left.isEmpty() || right.isEmpty()) return Difference.CONTENT_CHANGED
        val lhs = left.joinToString(" ")
        val rhs = right.joinToString(" ")
        if (digitRuns(lhs) != digitRuns(rhs)) return Difference.CONTENT_CHANGED
        return if (phoneticKey(lhs) == phoneticKey(rhs)) Difference.SPELLING_FIXED
        else Difference.CONTENT_CHANGED
    }

    private fun phoneticKey(text: String): String {
        var value = text.lowercase().filter(Char::isLetter)
        if (value.isEmpty()) return ""
        listOf(
            "sch" to "sk", "ph" to "f", "ck" to "k", "kn" to "n",
            "wr" to "r", "gn" to "n", "gh" to "", "wh" to "w",
        ).forEach { (from, to) -> value = value.replace(from, to) }
        val mapped = buildString {
            value.forEachIndexed { index, character ->
                when (character) {
                    'c' -> append(
                        if (value.getOrNull(index + 1)?.let { "eiy".contains(it) } == true) 's'
                        else 'k',
                    )
                    'q' -> append('k')
                    'x' -> append("ks")
                    'z' -> append('s')
                    'v' -> append('f')
                    'a', 'e', 'i', 'o', 'u', 'y' -> append('a')
                    'h', 'w' -> Unit
                    else -> append(character)
                }
            }
        }
        return buildString { mapped.forEach { if (lastOrNull() != it) append(it) } }
    }

    private fun alignedDifferences(left: List<String>, right: List<String>): List<Span> {
        val lengths = Array(left.size + 1) { IntArray(right.size + 1) }
        for (i in left.lastIndex downTo 0) for (j in right.lastIndex downTo 0) {
            lengths[i][j] = if (normalizeToken(left[i]) == normalizeToken(right[j])) {
                lengths[i + 1][j + 1] + 1
            } else {
                maxOf(lengths[i + 1][j], lengths[i][j + 1])
            }
        }
        val spans = mutableListOf<Span>()
        val pendingLeft = mutableListOf<String>()
        val pendingRight = mutableListOf<String>()
        fun flush() {
            if (pendingLeft.isEmpty() && pendingRight.isEmpty()) return
            spans.add(Span(pendingLeft.toList(), pendingRight.toList()))
            pendingLeft.clear(); pendingRight.clear()
        }
        var i = 0
        var j = 0
        while (i < left.size && j < right.size) {
            if (normalizeToken(left[i]) == normalizeToken(right[j])) {
                flush(); i++; j++
            } else if (lengths[i + 1][j] >= lengths[i][j + 1]) {
                pendingLeft.add(left[i++])
            } else {
                pendingRight.add(right[j++])
            }
        }
        while (i < left.size) pendingLeft.add(left[i++])
        while (j < right.size) pendingRight.add(right[j++])
        flush()
        return spans
    }

    private fun usableLearnedTerm(raw: String): String? {
        val candidate = raw.trim(*"\"“”‘’(),;:!?[]{} \t\r\n".toCharArray())
        val term = runCatching { normalize(candidate) }.getOrNull() ?: return null
        return term.takeIf {
            it.codePointCount(0, it.length) >= 3 && it.any(Char::isLetter) && it.none(Char::isDigit)
        }
    }

    private fun bestSpellingSubspan(left: List<String>, right: List<String>): String? {
        var best: Pair<Int, String>? = null
        for (leftStart in left.indices) for (rightStart in right.indices) {
            for (leftCount in 1..minOf(3, left.size - leftStart)) {
                for (rightCount in 1..minOf(3, right.size - rightStart)) {
                    val lhs = left.subList(leftStart, leftStart + leftCount)
                    val rhs = right.subList(rightStart, rightStart + rightCount)
                    if (classify(lhs, rhs) != Difference.SPELLING_FIXED) continue
                    val term = usableLearnedTerm(rhs.joinToString(" ")) ?: continue
                    val candidate = leftCount + rightCount to term
                    if (best == null || candidate.first > best!!.first) best = candidate
                }
            }
        }
        return best?.second
    }
}

File: windows/DoNotType.Core/PersonalDictionary.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace DoNotType.Core;

/// <summary>A bounded, explicit spelling reference supplied by the user.</summary>
public static partial class PersonalDictionary
{
    public const int MaxTerms = 100;
    public const int MaxCharactersPerTerm = 50;

    public sealed class ValidationException(string message) : Exception(message);

    public static string Normalize(string raw)
    {
        if (raw.IndexOfAny(['\r', '\n']) >= 0)
            throw new ValidationException("A dictionary entry must fit on one line.");
        var term = string.Join(' ', raw.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
        if (term.Length == 0) throw new ValidationException("Enter a word or phrase.");
        if (new StringInfo(term).LengthInTextElements > MaxCharactersPerTerm)
            throw new ValidationException($"Dictionary entries can be at most {MaxCharactersPerTerm} characters.");
        return term;
    }

    public static IReadOnlyList<string> Sanitize(IEnumerable<string>? raw)
    {
        var result = new List<string>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var value in raw ?? [])
        {
            string term;
            try { term = Normalize(value); }
            catch (ValidationException) { continue; }
            if (!seen.Add(term)) continue;
            result.Add(term);
            if (result.Count == MaxTerms) break;
        }
        return result;
    }

    public static IReadOnlyList<string> Add(string raw, IEnumerable<string> terms)
    {
        var current = Sanitize(terms).ToList();
        if (current.Count >= MaxTerms)
            throw new ValidationException($"The dictionary can contain at most {MaxTerms} entries.");
        var term = Normalize(raw);
        if (current.Contains(term, StringComparer.OrdinalIgnoreCase))
            throw new ValidationException($"“{term}” is already in the dictionary.");
        current.Add(term);
        return current;
    }

    public static IReadOnlyList<string> Replace(string original, string raw, IEnumerable<string> terms)
    {
        var current = Sanitize(terms).ToList();
        var index = current.IndexOf(original);
        if (index < 0) return current;
        var term = Normalize(raw);
        if (current.Where((_, offset) => offset != index)
            .Contains(term, StringComparer.OrdinalIgnoreCase))
            throw new ValidationException($"“{term}” is already in the dictionary.");
        current[index] = term;
        return current;
    }

    /// <summary>Reads Typeless-compatible UTF-8, one-column CSV (or plain lines).</summary>
    public static IReadOnlyList<string> EntriesFromCsv(byte[] data)
    {
        string text;
        try { text = new UTF8Encoding(false, true).GetString(data); }
        catch (DecoderFallbackException)
        {
            throw new ValidationException("The file is not UTF-8 text.");
        }
        text = text.TrimStart('\uFEFF').Replace("\r\n", "\n").Replace('\r', '\n');
        var result = new List<string>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var lines = text.Split('\n');
        for (var offset = 0; offset < lines.Length; offset++)
        {
            var line = lines[offset];
            if (string.IsNullOrWhiteSpace(line)) continue;
            var term = Normalize(CsvField(line, offset + 1));
            if (!seen.Add(term)) continue;
            result.Add(term);
            if (result.Count > MaxTerms)
                throw new ValidationException($"The dictionary can contain at most {MaxTerms} entries.");
        }
        return result;
    }

    public static IReadOnlyList<string> Import(IEnumerable<string> imported, IEnumerable<string> terms)
    {
        var result = Sanitize(terms).ToList();
        var seen = new HashSet<string>(result, StringComparer.OrdinalIgnoreCase);
        foreach (var raw in imported)
        {
            var term = Normalize(raw);
            if (!seen.Add(term)) continue;
            if (result.Count >= MaxTerms)
                throw new ValidationException($"The dictionary can contain at most {MaxTerms} entries.");
            result.Add(term);
        }
        return result;
    }

    public static string? ReferenceBlock(IEnumerable<string> raw)
    {
        var terms = Sanitize(raw);
        if (terms.Count == 0) return null;
        return """
            PERSONAL DICTIONARY — SPELLING REFERENCE ONLY, DO NOT TRANSCRIBE
            The user supplied the JSON strings below as possible spellings. Use an entry only when
            the same word or phrase is audible. The list is not evidence that an entry was spoken,
            and it never overrides clear audio. Digits, versions and quantities come from audio
            alone even when an entry contains a number.
            """ + "\n" + JsonSerializer.Serialize(terms) + "\n" +
            "END PERSONAL DICTIONARY. The audio is still the ONLY thing to transcribe.";
    }

    public static IReadOnlyList<string> Keyterms(
        IEnumerable<string> raw, int maxTerms, int maxCharactersPerTerm)
    {
        if (maxTerms <= 0) return [];
        return Sanitize(raw).Where(term =>
                !term.Any(char.IsDigit)
                && new StringInfo(term).LengthInTextElements <= maxCharactersPerTerm)
            .Take(maxTerms).ToList();
    }

    public static IReadOnlyList<string> MergeKeyterms(
        IEnumerable<string> dictionary, IEnumerable<string> derived,
        int maxTerms, int maxCharactersPerTerm)
    {
        var result = Keyterms(dictionary, maxTerms, maxCharactersPerTerm).ToList();
        var seen = new HashSet<string>(result, StringComparer.OrdinalIgnoreCase);
        foreach (var term in derived)
        {
            if (new StringInfo(term).LengthInTextElements > maxCharactersPerTerm || !seen.Add(term))
                continue;
            result.Add(term);
            if (result.Count == maxTerms) break;
        }
        return result;
    }

    /// <summary>Returns only stable spelling/capitalisation fixes from a post-insertion edit.</summary>
    public static IReadOnlyList<string> LearnedCandidates(string original, string corrected)
    {
        if (original == corrected) return [];
        var left = Tokenize(original);
        var right = Tokenize(corrected);
        var candidates = new List<string>();

        foreach (var difference in AlignedDifferences(left, right))
        {
            if (Classify(difference.Left, difference.Right) == Difference.SpellingFixed)
            {
                if (UsableLearnedTerm(string.Join(' ', difference.Right)) is { } term)
                    candidates.Add(term);
            }
            else
            {
                var subspan = BestSpellingSubspan(difference.Left, difference.Right);
                if (subspan is not null) candidates.Add(subspan);
            }
        }

        if (left.Count == right.Count)
        {
            for (var i = 0; i < left.Count; i++)
            {
                if (NormalForm(left[i]) == NormalForm(right[i]) && left[i] != right[i]
                    && string.Equals(left[i], right[i], StringComparison.OrdinalIgnoreCase)
                    && UsableLearnedTerm(right[i]) is { } term)
                    candidates.Add(term);
            }
        }
        return Sanitize(candidates);
    }

    private static string CsvField(string line, int lineNumber)
    {
        var trimmed = line.Trim();
        if (!trimmed.StartsWith('"'))
        {
            if (line.Contains(','))
                throw new ValidationException($"Line {lineNumber} has more than one CSV column. Use one entry per row.");
            return line;
        }

        var value = new StringBuilder();
        var index = 1;
        var closed = false;
        while (index < trimmed.Length)
        {
            if (trimmed[index] == '"')
            {
                if (index + 1 < trimmed.Length && trimmed[index + 1] == '"')
                {
                    value.Append('"'); index += 2; continue;
                }
                closed = true; index++; break;
            }
            value.Append(trimmed[index++]);
        }
        if (!closed)
            throw new ValidationException($"Line {lineNumber} has an unterminated or malformed quoted value.");
        var remainder = trimmed[index..].Trim();
        if (remainder.Length > 0)
        {
            if (remainder.StartsWith(','))
                throw new ValidationException($"Line {lineNumber} has more than one CSV column. Use one entry per row.");
            throw new ValidationException($"Line {lineNumber} has an unterminated or malformed quoted value.");
        }
        return value.ToString();
    }

    private enum Difference { SpellingFixed, ContentChanged }

    private static List<string> Tokenize(string text) =>
        WhitespaceRegex().Split(text.Trim()).Where(value => NormalForm(value).Length > 0).ToList();

    private static string NormalForm(string token) =>
        new(token.ToLowerInvariant().Where(char.IsLetterOrDigit).ToArray());

    private static Difference Classify(IReadOnlyList<string> left, IReadOnlyList<string> right)
    {
        if (left.Count == 0 || right.Count == 0) return Difference.ContentChanged;
        var lhs = string.Join(' ', left);
        var rhs = string.Join(' ', right);
        if (!DigitRuns(lhs).SequenceEqual(DigitRuns(rhs))) return Difference.ContentChanged;
        return PhoneticKey(lhs) == PhoneticKey(rhs) ? Difference.SpellingFixed : Difference.ContentChanged;
    }

    private static IEnumerable<string> DigitRuns(string value) =>
        DigitRegex().Matches(value).Select(match => match.Value);

    private static string PhoneticKey(string value)
    {
        var letters = new string(value.ToLowerInvariant().Where(char.IsLetter).ToArray());
        if (letters.Length == 0) return string.Empty;
        foreach (var (from, to) in new (string, string)[]
        {
            ("sch", "sk"), ("ph", "f"), ("ck", "k"), ("kn", "n"),
            ("wr", "r"), ("gn", "n"), ("gh", ""), ("wh", "w"),
        }) letters = letters.Replace(from, to, StringComparison.Ordinal);

        var mapped = new StringBuilder();
        for (var index = 0; index < letters.Length; index++)
        {
            var character = letters[index];
            switch (character)
            {
                case 'c': mapped.Append(index + 1 < letters.Length && "eiy".Contains(letters[index + 1]) ? 's' : 'k'); break;
                case 'q': mapped.Append('k'); break;
                case 'x': mapped.Append("ks"); break;
                case 'z': mapped.Append('s'); break;
                case 'v': mapped.Append('f'); break;
                case 'a' or 'e' or 'i' or 'o' or 'u' or 'y': mapped.Append('a'); break;
                case 'h' or 'w': break;
                default: mapped.Append(character); break;
            }
        }
        var collapsed = new StringBuilder();
        foreach (var character in mapped.ToString())
            if (collapsed.Length == 0 || collapsed[^1] != character) collapsed.Append(character);
        return collapsed.ToString();
    }

    private sealed record Span(IReadOnlyList<string> Left, IReadOnlyList<string> Right);

    private static IReadOnlyList<Span> AlignedDifferences(IReadOnlyList<string> left, IReadOnlyList<string> right)
    {
        var lengths = new int[left.Count + 1, right.Count + 1];
        for (var i = left.Count - 1; i >= 0; i--)
            for (var j = right.Count - 1; j >= 0; j--)
                lengths[i, j] = NormalForm(left[i]) == NormalForm(right[j])
                    ? lengths[i + 1, j + 1] + 1
                    : Math.Max(lengths[i + 1, j], lengths[i, j + 1]);

        var result = new List<Span>();
        var lhs = new List<string>();
        var rhs = new List<string>();
        void Flush()
        {
            if (lhs.Count == 0 && rhs.Count == 0) return;
            result.Add(new Span(lhs.ToList(), rhs.ToList())); lhs.Clear(); rhs.Clear();
        }

        var x = 0; var y = 0;
        while (x < left.Count || y < right.Count)
        {
            if (x < left.Count && y < right.Count && NormalForm(left[x]) == NormalForm(right[y]))
            { Flush(); x++; y++; }
            else if (x < left.Count && (y == right.Count || lengths[x + 1, y] >= lengths[x, y + 1]))
                lhs.Add(left[x++]);
            else if (y < right.Count) rhs.Add(right[y++]);
        }
        Flush();
        return result;
    }

    private static string? UsableLearnedTerm(string raw)
    {
        var candidate = raw.Trim("\"“”‘’(),;:!?[]{} \t\r\n".ToCharArray());
        string term;
        try { term = Normalize(candidate); }
        catch (ValidationException) { return null; }
        return new StringInfo(term).LengthInTextElements >= 3
            && term.Any(char.IsLetter) && !term.Any(char.IsDigit) ? term : null;
    }

    private static string? BestSpellingSubspan(IReadOnlyList<string> left, IReadOnlyList<string> right)
    {
        (int Coverage, string Term)? best = null;
        for (var ls = 0; ls < left.Count; ls++)
            for (var rs = 0; rs < right.Count; rs++)
                for (var lc = 1; lc <= Math.Min(3, left.Count - ls); lc++)
                    for (var rc = 1; rc <= Math.Min(3, right.Count - rs); rc++)
                    {
                        var lhs = left.Skip(ls).Take(lc).ToList();
                        var rhs = right.Skip(rs).Take(rc).ToList();
                        if (Classify(lhs, rhs) != Difference.SpellingFixed
                            || UsableLearnedTerm(string.Join(' ', rhs)) is not { } term) continue;
                        if (best is null || lc + rc > best.Value.Coverage) best = (lc + rc, term);
                    }
        return best?.Term;
    }

    [GeneratedRegex(@"\s+")]
    private static partial Regex WhitespaceRegex();
    [GeneratedRegex(@"\d+")]
    private static partial Regex DigitRegex();
}
